### PR TITLE
Refactor the logging module and clean some code

### PR DESCRIPTION
**Pull request by Zheaoli**
Signed-off-by: Zheao.Li <me@manjusaka.me>

### CODE REVIEW

File: pkg/logging/logging.go [hunk 1]
```diff
@@ -17,83 +17,47 @@
...
 //
 // Should be called only if argv1 == MagicArgv1.
 func Main(argsMap map[string]string) error {
-	fn, err := getLoggerFunc(argsMap)
+	if _, ok := argsMap[DriverType]; !ok {
+		return errors.New("the driverType is empty")
+	}
+	var driver Interface
+	// TODO: Impl more drivers such as local, journald, fluentd
+	if argsMap[DriverType] == "json-file" {
```
Comment (fahedouch) [inline pkg/logging/logging.go:48]: preparing to other drivers, it more humain friendly if we use a switch case here

File: pkg/logging/json_logger.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+		}
+		if maxFile < 1 {
+			return fmt.Errorf("max-file cannot be less than 1")
+		}
+	}
+	jsonLogger.maxFile = maxFile - 1
+	return nil
+}
+
+func (jsonLogger *JsonLogger) getLoggerFunc(argsMap map[string]string) (logging.LoggerFunc, error) {
```
Comment (junnplus) [inline pkg/logging/json_logger.go:70]: Why not just return `logging.LoggerFunc`?

File: pkg/logging/logging.go [hunk 3]
```diff
@@ -17,83 +17,55 @@
...
+	getLoggerFunc(argsMap map[string]string) (logging.LoggerFunc, error)
+}
+
 // Main is the entrypoint for the containerd runtime v2 logging plugin mode.
 //
 // Should be called only if argv1 == MagicArgv1.
 func Main(argsMap map[string]string) error {
-	fn, err := getLoggerFunc(argsMap)
+	if _, ok := argsMap[DriverType]; !ok {
+		return errors.New("the driverType is empty")
```
Comment (fahedouch) [inline pkg/logging/logging.go:48]: no need for this;  driver in never, it has a default value `json-file`

File: pkg/logging/logging.go [hunk 4]
```diff
@@ -17,83 +17,55 @@
...
+	driverType := argsMap[DriverType]
+	switch driverType {
+	case jsonFile:
+		driver = &JsonLogger{}
+	default:
+		driver = nil
+	}
+	if driver == nil {
+		return fmt.Errorf("%s is not yet supported", argsMap[DriverType])
+	}
```
[the remark this answers is not in the document]
Comment (fahedouch) [reply pkg/logging/logging.go:61]: no need for error message, we can default to nothing : https://github.com/containerd/nerdctl/blob/70bcdd06417a652be535a68301ce1dc2698c9a30/cmd/nerdctl/run.go#L770

File: pkg/logging/logging.go (hunk 3 above)
Comment (Zheaoli) [reply pkg/logging/logging.go:48]: OK

File: cmd/nerdctl/run.go [hunk 5]
```diff
@@ -771,6 +771,7 @@ func generateLogURI(dataStore, logDriver string, logOptMap map[string]string) (*
 	}
 	args := map[string]string{
 		logging.MagicArgv1: dataStore,
+		logging.DriverType: logDriver,
```
Comment (AkihiroSuda) [inline cmd/nerdctl/run.go:774]: I don't think this is correct.

We have to ensure that MagicArgv1 is always argv[1], but this map may make it argv[3]

File: pkg/logging/logging.go [hunk 6]
```diff
@@ -17,83 +17,48 @@
...
+	driverType := argsMap[DriverType]
+	switch driverType {
+	case jsonFile:
+		driver = &JsonLogger{}
+	default:
+		return fmt.Errorf("%s is not yet supported", argsMap[DriverType])
+	}
+	if driver == nil {
+		return fmt.Errorf("%s is not yet supported", argsMap[DriverType])
+	}
```
Comment (junnplus) [inline pkg/logging/logging.go:57]: Remove it.

File: pkg/logging/logging.go [hunk 7]
```diff
@@ -17,83 +17,48 @@
...
-	fn, err := getLoggerFunc(argsMap)
-	if err != nil {
+	var driver NerdCTLLoggingInterface
+	// TODO: Impl more drivers such as local, journald, fluentd
+	driverType := argsMap[DriverType]
+	switch driverType {
+	case jsonFile:
+		driver = &JsonLogger{}
+	default:
+		return fmt.Errorf("%s is not yet supported", argsMap[DriverType])
```
Comment (junnplus) [inline pkg/logging/logging.go:53]: Like @fahedouch said, we don't need it, just return.

**Comment by Zheaoli** [comment]
@AkihiroSuda 

@fahedouch and I want to enhance the log driver, see #1039 

In the old version of the code, we hard-coded some `json-file` only logic in the logging module, I think it's not a good choice if we want to support multiple logging-driver. So I refactor the hard-coded code to Interface

**Comment by fahedouch** [comment]
@AkihiroSuda, I agree with @Zheaoli , this refacto will help us handling this list of logging driver  https://github.com/containerd/nerdctl/issues/1039

File: cmd/nerdctl/run.go (hunk 5 above)
Comment (fahedouch) [reply cmd/nerdctl/run.go]: @AkihiroSuda  not really, I have already experimented with a change of position of the args in this PR https://github.com/containerd/nerdctl/pull/959. 

the actual code
```
	args := map[string]string{
		logging.MagicArgv1: dataStore,
	}
	for k, v := range logOptMap {
		args[k] = v
	}
```
does not guarantee a fixed order of arguments
for this reason  the args parsing is refactored in this way : https://github.com/containerd/nerdctl/pull/959/files#diff-055f400ea02598bd9b9ef894b5ffc400954e88ad71cc43a31f421222af4da6c7R81

**Review by junnplus — APPROVED** [review]
LGTM

File: pkg/logging/logging.go [hunk 8]
```diff
@@ -17,83 +17,43 @@
...
 	MaxSize    = "max-size"
 	MaxFile    = "max-file"
+	DriverType = "logging-driver"
 )
 
+const (
+	jsonFile = "json-file"
+)
+
+type NerdCTLLoggingInterface interface {
```
Comment (fahedouch) [inline pkg/logging/logging.go:35]: I suggest `type NerdctlLogger interface`.

**Review by fahedouch — APPROVED** [review]

**Comment by AkihiroSuda** [comment]
I think this kind of refactoring should be done when you actually implement another driver.
Hard to review whether the design is correct otherwise.

**Comment by Zheaoli** [comment]
> I think this kind of refactoring should be done when you actually implement another driver.
> 
> Hard to review whether the design is correct otherwise.
> 
> 
> 
> 

@fahedouch has worked on journal log driver base on this PR. FYI https://github.com/containerd/nerdctl/issues/1039#issuecomment-1123675221

**Comment by fahedouch** [comment]
@Zheaoli no problem, I rebased your refacto on my branch, I will try to push a PR as soon as possible. Waiting to merge https://github.com/containerd/nerdctl/pull/1056 this PR too

**Pull request closed without merging**